File: server-node/reset_password.js
#!/usr/bin/env node
'use strict';
/**
 * reset_password.js — reset the Relay Controller dashboard password
 * ───────────────────────────────────────────────────────────────────
 * Run this on the machine that hosts server.js (it edits password.json
 * next to it, so it must stay in the same folder as auth.js).
 *
 * Usage:
 *   node reset_password.js
 *       Prompts for a new password twice (hidden input) and saves it.
 *
 *   node reset_password.js --password "NewPassword123"
 *       Non-interactive — sets the password directly (careful: this may
 *       end up in your shell history / process list).
 *
 * The server does not need to be running for this to work, and if it IS
 * running you do not need to restart it — the password hash is re-read
 * from disk on every login attempt, so the new password takes effect
 * immediately.
 */
const readline = require('readline');
const auth = require('./auth');

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--password' && argv[i + 1] !== undefined) {
      out.password = argv[i + 1];
      i++;
    }
  }
  return out;
}

/** Prompt for input with the typed characters hidden (masked as *). */
function promptHidden(question) {
  return new Promise((resolve) => {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    // Hide echoed input by intercepting the underlying writer.
    const originalWrite = rl._writeToOutput;
    rl._writeToOutput = function hiddenWrite(stringToWrite) {
      if (stringToWrite.charCodeAt(0) === 13 || stringToWrite.includes('\n')) {
        originalWrite.call(rl, stringToWrite);
      } else {
        originalWrite.call(rl, '*');
      }
    };
    rl.question(question, (answer) => {
      rl.close();
      process.stdout.write('\n');
      resolve(answer);
    });
  });
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  let newPassword = args.password;

  if (!newPassword) {
    console.log('Reset Relay Controller dashboard password');
    console.log('-------------------------------------------');
    const p1 = await promptHidden('New password: ');
    const p2 = await promptHidden('Confirm password: ');
    if (p1 !== p2) {
      console.error('Error: passwords do not match.');
      process.exitCode = 1;
      return;
    }
    newPassword = p1;
  }

  try {
    auth.setPassword(newPassword);
  } catch (err) {
    console.error(`Error: ${err.message}`);
    process.exitCode = 1;
    return;
  }

  console.log(`Password updated successfully. (${auth.PASSWORD_FILE})`);
  console.log('Log in to the dashboard with the new password — no server restart needed.');
}

main();

File: defunct/server/reset_password.py
#!/usr/bin/env python3
"""
reset_password.py — reset the Relay Controller dashboard password
───────────────────────────────────────────────────────────────────
Run this on the machine that hosts server.py (it edits password.json
in the same folder, so it must be run from here, or from anywhere as
long as this script stays next to auth.py).

Usage:
    python reset_password.py
        Prompts for a new password twice (hidden input) and saves it.

    python reset_password.py --password "NewPassword123"
        Non-interactive — sets the password directly (careful: this may
        end up in your shell history / process list).

The dashboard server does not need to be running for this to work, but
if it IS running, restart it afterwards so nothing gets confused (the
password hash is re-read from disk on every login attempt, so this is
actually not required — the new password takes effect immediately).
"""
import argparse
import getpass
import sys

import auth  # local module, must stay in the same folder


def main() -> int:
    parser = argparse.ArgumentParser(description="Reset the Relay Controller dashboard password.")
    parser.add_argument(
        "--password",
        help="New password (non-interactive). If omitted, you'll be prompted securely.",
    )
    args = parser.parse_args()

    if args.password:
        new_password = args.password
    else:
        print("Reset Relay Controller dashboard password")
        print("-------------------------------------------")
        new_password = getpass.getpass("New password: ")
        confirm = getpass.getpass("Confirm password: ")
        if new_password != confirm:
            print("Error: passwords do not match.")
            return 1

    try:
        auth.set_password(new_password)
    except ValueError as e:
        print(f"Error: {e}")
        return 1

    print(f"Password updated successfully. ({auth.PASSWORD_FILE})")
    print("Log in to the dashboard with the new password — no server restart needed.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
